xbmc/video/VideoInfoScanner.cpp

```diff
@@ -565,6 +568,117 @@ CVideoInfoScanner::~CVideoInfoScanner()
+    CFileItemList items;
+    CDirectory::GetDirectory(msifPath, items, "", DIR_FLAG_DEFAULTS);
+
+    // Remove non-subfolders, folders flagged as no media and excluded folders
+    items.erase(std::remove_if(items.begin(), items.end(),
+                               [&regexps](const CFileItemPtr& item)
+                               {
+                                 return !item->IsFolder() || HasNoMedia(item->GetPath()) ||
+                                        CUtil::ExcludeFileOrFolder(item->GetPath(), regexps);
+                               }),
+                items.end());
+
+    // No folders to process
+    if (items.IsEmpty())
+    {
+      CLog::LogF(LOGDEBUG, "No directories to scan in the Movie Set Information Folder");
+      return;
+    }
+
+    if (m_handle)
+      m_handle->SetTitle(g_localizeStrings.Get(408)); // 'Updating sets'
+
+    for (int i = 0; const auto& item : items)
+    {
+      // Folder name is the set title
+      std::string path{item->GetPath()};
+      std::string setTitle{path};
+      URIUtils::RemoveSlashAtEnd(setTitle);
+      setTitle = URIUtils::GetFileName(setTitle);
+
+      i++;
+      if (m_handle)
+      {
+        m_handle->SetPercentage(static_cast<float>(i) * 100.0f / static_cast<float>(items.Size()));
+        m_handle->SetText(setTitle);
+      }
+
+      CFileItemList nfos;
+      CDirectory::GetDirectory(path, nfos, "*.nfo", DIR_FLAG_DEFAULTS);
+
+      // Get hash for the folder
+      // (Not using fast hash as file changes are not reflected in the fast hash)
+      std::string hash;
+      std::string dbHash;
+      GetPathHash(nfos, hash);
+      m_database.GetPathHash(path, dbHash);
+      if (StringUtils::EqualsNoCase(hash, dbHash))
+      {
+        CLog::LogF(LOGDEBUG, "Skipping directory {} as no change", CURL::GetRedacted(path));
+        continue; // Hashes match - no need to process anything
+      }
+
+      // Save hash
+      m_database.SetPathHash(path, hash);
+
+      // Remove folders, non-NFO files and excluded files
+      nfos.erase(std::remove_if(nfos.begin(), nfos.end(),
+                                [&regexps](const CFileItemPtr& item)
+                                {
+                                  return item->IsFolder() || !item->IsNFO() ||
+                                         CUtil::ExcludeFileOrFolder(item->GetPath(), regexps);
+                                }),
+                 nfos.end());
+
+      // Should only be one NFO file in the folder
+      if (nfos.Size() != 1)
+      {
+        CLog::LogF(LOGDEBUG, "Skipping directory {} as multiple nfos", CURL::GetRedacted(path));
+        continue;
+      }
+
+      // Set.nfo is an override set, so get existing set
+      CSetInfoTag setTag;
+      if (!m_database.GetSetInfo(setTitle, setTag))
+      {
+        CLog::LogF(LOGDEBUG, "Skipping directory {} as there is no existing set {}",
+                   CURL::GetRedacted(path), setTitle);
+        continue; // No existing set
+      }
```

Comment:
Would it be a problem to have a set created in the db "in advance" by a MSIF folder with .nfo/art?